### DISCUSSION TITLE
Can pudb use python as it configure file?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [inducer/pudb#584] [Ideas]
Many python develop tools use python as its config file, such as python use
`$PYTHONSTARTUP` which is a python file, ptpython use
`~/.config/ptpython/config.py`, ipython use
`~/.ipython/profile_default/ipython_config.py`. How about let pudb use python to
config itself? I think it can provide more flexibility than dosini file,
especially define of hot key. A
referable example is <https://github.com/prompt-toolkit/ptpython/blob/master/examples/ptpython_config/config.py>

```python
def configure(repl):
    # define some values of settings
    repl.show_docstring = False
    ...

    # define new hotkey
    @repl.add_key_binding("c-b")
    def _(event):
        ...
```

I think all users of pudb should know how to write python (pudb is python
debugger) so the situation that user don't know how to write the config shouldn't exist.

**Comment by inducer** [comment]
It's a trade-off. We programmatically write to the config file (when changes to the preferences dialog are made), and I'm sure it would be sort of unpopular if we obliterated code in the config file every time a change is made from the UI. It's not *unheard of* for config files to have both machine-written and code sections (Emacs's config file comes to mind), but it's sort of more engineering than I'd like to do.

It's always possible to make it a separate file, but that's also kind of confusing.

In short, don't know.

**Reply by Freed-Wu** [reply]
![screen-2023-01-14-03-26-44](https://user-images.githubusercontent.com/32936898/212402413-8ad4146d-9df6-40fb-86fd-321451dfc13d.jpg)
As a complement, ptpython also support config from TUI, however, the change is temporary, that is if users restart ptpython, ptpython still read `~/.config/ptpython/config.py`, the change will not be saved. Perhaps it is a simple solution.

**Reply by inducer** [reply]
I remember stumbling on that, and I found it confusing. It would also be a change of behavior vs what's happening now.

**Comment by asmeurer** [comment]
What's the actual use-case for a dynamic configuration? You want to define hotkeys that depend on other values?

**Reply by Freed-Wu** [reply]
Oh, I hope to customize some advanced hotkeys, such as jump to next position of class method definition.
